### PR TITLE
Weight Standardization method

### PR DESCRIPTION
**Pull request by Landanjs**
Composer algorithm implementation for [weight standardization](https://arxiv.org/abs/1903.10520). 

Positive results when training ResNet-50 on ImageNet and training DeepLabv3+ on ADE20k. Weight standardization on DeepLabv3+ only worked when ignoring the last 15 convolutional layers (i.e. the DeepLabv3+ head). Still not sure why standardizing these layers is harmful...

I was unable to get alternative normalization layers like LayerNorm and GroupNorm to work with weight standardization. Throughput was cut in half and convergence wasn't as good for ResNet-50 on ImageNet. It seemed like training was unstable since there were loss spikes, but ensure why this is happening in our benchmarks, but not in the other scenarios weight standardization is used (i.e. BiT).

[WandB runs](https://wandb.ai/mosaic-ml/weight-standardization), then group by "Name". There are a lot of runs, but the important ones are:
- "dlv3p-ade20k-dice-baseline-nopretrain" (DeepLabv3+ baseline)
- "dlv3p-ade20k-dice-ws_ignored15-nopretrain" (DeepLabv3+ weight standardization and ignore last 15 layers)
- "r50-i1k-baseline" (ResNet50 baseline)
- "r50-i1k-weight-standard" (ResNet50 weight standardization)

### CODE REVIEW

**Comment by eracah** [comment]
nice use of the torch `parametrize` tool!

**Comment by Landanjs** [comment]
@dblalock @dskhudia it added y'all by default, feel free to ignore this or give a fake review. I'll get Austin or Cory to review this

File: composer/algorithms/weight_standardization/README.md
Comment (coryMosaicML) [inline composer/algorithms/weight_standardization/README.md:102]: Do we see this improvement when using pretrained weights, or only when training from scratch?

File: composer/algorithms/weight_standardization/weight_standardization.py
Comment (coryMosaicML) [inline composer/algorithms/weight_standardization/weight_standardization.py:25]: What do you think about adding an option here to account for the fan in? Not quite full scaled weight standardization, but a step closer. Default behavior would still be vanilla weight standardization, but scaling by `1/sqrt(fan_in)` here might be helpful, but not different enough to warrant being it's own algorithm.

File: composer/algorithms/weight_standardization/weight_standardization.py
Comment (coryMosaicML) [inline composer/algorithms/weight_standardization/weight_standardization.py:46]: Do you know why this might fail? If so, maybe include a bit about it in the warning.

**Review by coryMosaicML — APPROVED** [review]
Overall looks good! One minor question and suggested an optional improvement.

File: composer/algorithms/weight_standardization/README.md
Comment (Landanjs) [reply composer/algorithms/weight_standardization/README.md:106]: Only training from scratch since I haven't tried using pre-trained weights trained with weight standardization. I will update the doc!

File: composer/algorithms/weight_standardization/weight_standardization.py
Comment (Landanjs) [reply composer/algorithms/weight_standardization/weight_standardization.py:46]: This was actually failing with our BERT model, but I did not dig into it 😅

Comment (Landanjs) [reply composer/algorithms/weight_standardization/weight_standardization.py:46]: It looks like this was failing due to unsupported operations in the forward function. Specifically, there is a for-loop with an if-statement in the `forward()` of our huggingface integration which is not support by `symbolic_trace()`. I added more details to the warning. Let me know if it is sufficient!

**Comment by Landanjs** [comment]
@dskhudia or @dblalock looks like I can't merge without a review from one of y'all. Would it be okay to get a stamp approval? Thanks!

**Review by dskhudia — APPROVED** [review]
rubber stamp

File: composer/algorithms/weight_standardization/weight_standardization.py
Comment (dblalock) [inline composer/algorithms/weight_standardization/weight_standardization.py:51]:
```suggestion
                Module could not be symbolically traced likely due to logic in forward() which is not traceable. Modules
```

**Review by dblalock — APPROVED** [review]
Basically rubber stamp as requested. Just found a couple typos / slight opportunities for clarification

File: composer/algorithms/weight_standardization/README.md
Comment (Landanjs) [reply composer/algorithms/weight_standardization/README.md:106]: Ambitious wording on my end, I'll change it to "similar accuracy"

**Pull request merged**
